File: sdk/ts/src/audit.ts
/**
 * Audit log resource (`/v1/audit`).
 *
 * Admin-only on the gateway side; calls will 403 unless the caller's
 * JWT carries `role=admin`. Filter on tenant/action/time-range; the
 * server enforces a 5 000-row hard cap regardless of the requested
 * limit.
 */

import type { VelocityClient } from './client.js'

export type AuditEvent = {
  event_id: string
  tenant_id: string
  subject: string
  role: string
  source: string
  action: string
  resource_type: string
  resource_id: string
  outcome: 'allow' | 'deny' | 'error'
  status_code: number
  occurred_at_ns: number
  remote_ip: string
  request_id: string
  chain_hash?: string
  meta?: Record<string, unknown>
}

export type AuditQuery = {
  tenant?: string
  action?: string
  since?: Date
  until?: Date
  limit?: number
}

export class AuditResource {
  constructor(private readonly client: VelocityClient) {}

  async query(q: AuditQuery = {}): Promise<AuditEvent[]> {
    const params = new URLSearchParams()
    if (q.tenant) params.set('tenant', q.tenant)
    if (q.action) params.set('action', q.action)
    if (q.since)  params.set('since',  q.since.toISOString())
    if (q.until)  params.set('until',  q.until.toISOString())
    if (q.limit)  params.set('limit',  String(q.limit))
    const resp = await this.client.request<{ events: AuditEvent[]; count: number }>(
      'GET', `/v1/audit?${params.toString()}`,
    )
    return resp.events
  }
}

File: sdk/go/audit.go
package velocity

import (
	"context"
	"net/url"
	"strconv"
	"time"
)

type AuditEvent struct {
	EventID      string         `json:"event_id"`
	TenantID     string         `json:"tenant_id"`
	Subject      string         `json:"subject"`
	Role         string         `json:"role"`
	Source       string         `json:"source"`
	Action       string         `json:"action"`
	ResourceType string         `json:"resource_type"`
	ResourceID   string         `json:"resource_id"`
	Outcome      string         `json:"outcome"`
	StatusCode   int            `json:"status_code"`
	OccurredAtNs int64          `json:"occurred_at_ns"`
	RemoteIP     string         `json:"remote_ip"`
	RequestID    string         `json:"request_id"`
	ChainHash    string         `json:"chain_hash"`
	Meta         map[string]any `json:"meta,omitempty"`
}

type AuditQuery struct {
	TenantID string
	Action   string
	Since    time.Time
	Until    time.Time
	Limit    int
}

type AuditResource struct{ c *Client }

func (r *AuditResource) Query(ctx context.Context, q *AuditQuery) ([]AuditEvent, error) {
	v := url.Values{}
	if q.TenantID != "" {
		v.Set("tenant", q.TenantID)
	}
	if q.Action != "" {
		v.Set("action", q.Action)
	}
	if !q.Since.IsZero() {
		v.Set("since", q.Since.Format(time.RFC3339))
	}
	if !q.Until.IsZero() {
		v.Set("until", q.Until.Format(time.RFC3339))
	}
	if q.Limit > 0 {
		v.Set("limit", strconv.Itoa(q.Limit))
	}
	var resp struct {
		Events []AuditEvent `json:"events"`
		Count  int          `json:"count"`
	}
	if err := r.c.do(ctx, "GET", "/v1/audit?"+v.Encode(), nil, &resp); err != nil {
		return nil, err
	}
	return resp.Events, nil
}
